Panama Withholding

hbto opened the pull request:
## Video Example:
https://drive.google.com/file/d/0B1w9gokcJL9hWHlNRDMzajFyN2M/view

## Todo List:
- [x] Mark Partners as Withholding Agent.
- [x] Field on Partner, Withholding Subject: To Discuss @nhomar, for reporting purpose it needs to be an integer, for business process needs to be a value, do we use a new record for this or do we wire the value to the integer value in Business Process.
- [x] Add wh_move_id on Invoice
- [x] Add server action to `open` node on Invoice Workflow to execute ITBMS Withholding
- [x] Withholding is only to Apply on Customer Invoice and Customer Refund - 9318d69
- [x] Fix Debit - Credit direction on Journal Items. - 9318d69
- [x] Reusing method compute_invoice_totals() on invoice. - 6f82f78
- [x] Omit Invoices from withholding, for example for petty cash. - 54638d8
- [x] When Withholding does not apply change the value from zero (current value) to 'na' because zero is not taken as an option to fill the field. - fc518da
- [x] Retrieve Accounting Configuration from Company and validate. - 19e09a8
- [x] Compute Withholding amount on each tax line base on Withholding Subject value on invoice 2ef3b5c
- [x] Withholding Agent Entitlement on invoice -    4245979
- [x] Marcar al Partner con campo para Agente de retención.
- [x] Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Sales Order - c21d2d4
- [x] Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Picking - d3cc67b
- [x] Taking away dependency on l10n_pa_reports - c866be3
- [x] Switching dependency toward l10n_pa_withholding from l10n_pa_reports - f9d8ba3
- [x] Add sale_stock to dependencies - 9934518
- [x] On Onchange Partner in Sales Order bring the Fiscal Information from AccountingPartner. - 8d64c37
- [x] Add Demo Data to be used in UnitTesting - 7f0180a1 - 20005fc - 33c81a9
- [x] UnitTesting Module - 7f0180a - 069d0d6 - 9fa03dc - 4987324 - 33c81a9 - f68d7a1 - 531f60f
- [x] Abstracting all duplicated fields in a common class/model - 7ad5250
- [x] Fiscal Info should only be available from Accounting Partner - 1419f48
- [x] Propagate field values to refund (Fiscal Info) when creating refund - de96186
- [x] Reconciliation Process, stitching together move_id on Invoice with wh_move_id from withholding.
- Without previous reconciliations on Invoice. [DONE]
- With previous reconciliatioin on Invoice. [DONE]
- [x] Adding Journal at company Settings to use when Withholding Taxes - 68c9a4f
- [x] Module Translation into Spanish -     4e6aaf0
- [x] There shall be some way to mark taxes as withholdable - ce45b8e
- [x] Process for cancellation of invoice - 80e9752
- [x] Keep track of wh_move_id.name (store it in invoice) to reuse it after cancellation - 6753199
- [x] Withholding is be created when using Withholding Subject in ('5', '6') - bae60c4 - da63cfd
- [ ] Field in account.move.line for afterwards Withholding Report purposes  -

## Other Pending Items
- [ ] Review Withholding Case Scenarios - Revisar Casos de retenciones.

hbto committed:
[ADD] Module for withholding taxes

nhomar commented on l10n_pa_withholding/__openerp__.py:
```diff
@@ -0,0 +1,24 @@
+    "website": "http://www.vauxoo.com/",
+    "license": "AGPL-3",
+    "depends": [
+            "l10n_pa_reports",
```
Why depend of reports? I was expecting this as a procedural modules only, in all case the one that should depend of this one is the l10n_pa_report

hbto replied:
Preliminar proposal not set in stone

hbto replied:
@nhomar 
Already fixed @ c866be3

nhomar commented on l10n_pa_withholding/model/account_invoice.py:
```diff
@@ -0,0 +1,47 @@
+    @api.model
+    def wh_move_line_get_item(self, line):
+        sign = 1
+        if line.invoice_id.type in ('out_invoice', 'in_invoice'):
```
this methodology is only valid for out_invoice and out_refund not for in_invoice that case has not been discussed yet.

hbto replied:
then just need to override the offending code

nhomar commented on l10n_pa_withholding/model/account_invoice.py:
```diff
@@ -0,0 +1,47 @@
+
+    @api.model
+    def wh_move_line_get(self):
+        """Creates Withholding for taxes in invoice"""
```
THe agreement was create an extra account move, not an extra tax.line, why are you coding an extra tax.line?

hbto replied:
As I have mentioned before to you privately, _there are two commits_ both are to be reviewed widen your sight

nhomar commented:
I can't see:
- Flag in partner to manage the withholding process.
- Process is not what we discussed, we discussed an extra account move not an extra tag.

Please fix.

hbto committed:
[IMP]Mark Partner with field for ITBMS Withholding Agent

hbto committed:
[FIX] File renaming

hbto committed:
[FIX] Fixing View inheritance and space arrangement on form

hbto committed:
[IMP] Adding Accounting Information for ITBMS on Company

hbto commented:
- [ ] Field on Partner, Withholding Subject: To Discuss @nhomar, for reporting purpose it needs to be an integer, for business process needs to be a float, that ranges from 1 to 100, do we use a new record for this or do we wire the value to the integer value in Business Process.

hbto committed:
[FIX] Taking away duplicated view

nhomar commented:
On Tue, Oct 25, 2016 at 3:47 PM, Hbto [ Vauxoo ] notifications@github.com
wrote:

> - Field on Partner, Withholding Subject: To Discuss @nhomar
>   https://github.com/nhomar, for reporting purpose it needs to be an
>   integer, for business process needs to be a value, do we use a new record
>   for this or do we wire the value to the integer value in Business Process.
> 
> It should be a selection with possible Values:

No Withhold
Percentage 1
Percentage 2
..
...
....
Percentage N

## 

Nhomar Hernandez
CEO Vauxoo.
Site: http://vauxoo.com
Twitter: @nhomar
Blog: http://nhomar.com
Github User: https://github.com/nhomar
Odoo Gold Partner
Skype: nhomar00 (Envia mail previo no lo superviso siempre).
HangOut: nhomar@vauxoo.com
Móvil Venezuela:
+58 4144110269
Móvil México:
+52 1 4773933942

hbto committed:
[IMP] Added Withholding Subject to Partner, Sale & Invoice Models & Views

hbto committed:
[IMP] Add wh_move_id on Invoice

hbto committed:
[IMP] Add Server Action to `open` node on Invoice Workflow to execute ITBMS Withholding

hbto committed:
[IMP] Withholding is only to Apply on Customer Invoice and Customer
Refund. And only to meaningful taxes to withhold.

hbto committed:
[IMP] Reusing method compute_invoice_totals() on invoice.

hbto committed:
[IMP] Omit Invoices from withholding, for example for petty cash

hbto committed:
[FIX] When Withholding does not apply change the value from zero
(current value) to 'na' because zero is not taken as an option to fill
the field.

hbto committed:
[IMP] Retrieve Accounting Configuration from Company and validate.

hbto committed:
[FIX] Wrong name on field for accounting configuration on company

hbto committed:
[FIX] First elements in selection field were set as strings

hbto committed:
[IMP] Compute Withholding amount on each tax line base on Withholding
Subject value on invoice

hbto committed:
[IMP] Withholding Agent Entitlement on invoice

hbto committed:
[IMP] Propagate Withholding Subject from Sale Order to Invoice, when
invoicing from Sales Order

hbto commented:
@nhomar 
Do we need to propagate info from SaleOrder to Invoice through path
SaleOrder > Picking > Invoice?

hbto committed:
[IMP] Propagate Withholding Subject from Sale Order to Invoice, when
invoicing from Picking

hbto commented:
@nhomar 

> Do we need to propagate info from SaleOrder to Invoice through path
> SaleOrder > Picking > Invoice?

Already done here:

Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Picking - d3cc67b

hbto committed:
[FIX] Missing dependency on manifest

nhomar commented on l10n_pa_withholding/__openerp__.py:
```diff
@@ -10,6 +10,7 @@
     "depends": [
             "l10n_pa_reports",
             "sale",
+            "stock",
```
can you add sale_stock to dependencies pliz

hbto replied:
@nhomar 
ok, got it

hbto replied:
@nhomar Done at 9934518

hbto committed:
[IMP] Taking away dependency on l10n_pa_reports

hbto committed:
[FIX] Switching dependency toward l10n_pa_withholding from l10n_pa_reports

hbto committed:
[FIX] l10n_pa_withholding depends on vat_partitioned

hbto committed:
[FIX] Add sale_stock to dependencies

hbto committed:
[IMP] On Onchange Partner in Sales Order bring the Fiscal Information
from AccountingPartner.

nhomar commented:
Según el plan:
- Campo en Aml de objeto de retención para efectos de reportes posteriores?

Puede ser un diario no? o un tipo de diario?

hbto committed:
[IMP] UnitTest: - Propagating Fiscal Info from Sales Order towards
Invoice via direct invoicing from Sales Order

hbto committed:
[IMP] UnitTest: - Testing Fiscal Info when there are no partner in Sale Order

hbto committed:
[IMP] UnitTest: - Propagating Fiscal Info from Sales Order towards
Invoice via direct invoicing from Picking

hbto committed:
[IMP] A straightforward way of doing things

hbto committed:
[IMP] Added a method to create invoice from sales order on UnitTest class

hbto committed:
[IMP] UnitTesting: Testing witholding an invoice without taxes

hbto committed:
[IMP] Added Demo Data For Taxes

hbto committed:
[IMP] UnitTesting: Test withholding in invoice with taxes but wh_agent_itbms=False

hbto committed:
[IMP] UnitTesting: Test Withholding in exempt taxed invoice and wh_agent_itbms=True

hbto committed:
[IMP] UnitTesting: Test withholding in invoice with taxes and wh_agent_itbms=True

hbto committed:
[IMP] Not really an useful statement

hbto committed:
[IMP] Taking away condition from the code

hbto committed:
[FIX] Taking away unused key to avoid warning on runbot

hbto committed:
[IMP] UnitTesting: Accounting information on Company

hbto committed:
[IMP] UnitTesting: Test withholding in invoice with taxes and
wh_agent_itbms=True No Withholding Subject set in the Invoice

hbto committed:
[IMP] UnitTesting: Testing Withholding on Refund and Proper values on Journal Entries

hbto committed:
[IMP] Abstracting all repeated fields a common Class

hbto committed:
[IMP] Fiscal Info should only be available from Accounting Partner

hbto committed:
[IMP] Numbering Test so they are executed in the intended order

hbto committed:
[IMP] Propagate field values to refund (Fiscal Info) when creating refund

hbto commented:
@nhomar 

Following Items are pending:
- [x] Module Translation into Spanish (DOING)
- [x] Field in account.move.line for afterwards Withholding Report purposes  - Campo en Aml de objeto de retención para efectos de reportes posteriores?
  Can we meet to further discuss this statement of yours
  
  > Puede ser un diario no? o un tipo de diario?
  > I have some doubts
- [x] Will we use the same journal than that of the invoice + a String for Journal Entry Naming or a new Journal?
  Do you have any idea or can we cross this item, getting rid of it.
- [ ] How to proceed on cancellation of invoice?
  What would you suggest?

hbto committed:
[IMP] Reconciliation Process, stitching together move_id on Invoice with
wh_move_id from withholding.

hbto committed:
[IMP] UnitTesting: Reconciliation Process

hbto committed:
[FIX] Not useful condition on list comprehension for wh_reconciliation

hbto committed:
[IMP] UnitTesting: Testing case when there is no change on Customer Receivable
though there are withholding documents

hbto committed:
[IMP] Consider payments in partially paid invoice when reconciling withholding

hbto committed:
[IMP] Adding Journal field on company to be used by Withholding Process

hbto committed:
[IMP] Getting rid of duplicate/unneeded code

hbto committed:
[IMP] Changing API to onchange method on SalesOrder

hbto commented:
@nhomar Could you review and provide further guidance.

Regards

hbto committed:
[IMP] Added Translation into Spanish

hbto committed:
[IMP] Added a new model to hold data where withholding will be fetched

hbto committed:
[PYLINT]

hbto committed:
[IMP] Tax is checked to be withheld

nhomar commented on l10n_pa_withholding/model/account_invoice.py:
```diff
@@ -191,3 +201,90 @@ def wihholding_reconciliation(self):
+class AccountInvoiceTaxWh(models.Model):
+    """Invoice Tax Withholding"""
+    _name = 'account.invoice.tax.wh'
+    _inherit = 'account.invoice.tax'
```
don't forget _rec_name

hbto replied:
There's no need for `_rec_name` I am prototyping from `account.invoice.tax` which already has `name`.

Or should I?
<img width="868" alt="models" src="https://cloud.githubusercontent.com/assets/7598010/19878599/1af6790e-9fc1-11e6-8ff6-6ca0a77836df.png">

hbto committed:
[IMP] Computing wh_amount inside account.invoice.tax.wh model fields

hbto committed:
[FIX] Demo data for tax must be withholdable

hbto committed:
[FIX] Fixing data in UnitTest

hbto committed:
[IMP] Added a new line to Sales Order to run through the code

hbto committed:
[FIX] If there are `wh_tax_line` records there are ait elements

hbto committed:
[IMP] Cancellation Process for Withholding Implemented

hbto committed:
[IMP] Squeezing the code a little bit

hbto committed:
[IMP] Adding Withholding name to invoice to reuse it when cancelling invoice

hbto committed:
[IMP] Increasing coverage on code by extending test cases

hbto committed:
[FIX] Conditional statements are senseless

hbto committed:
[IMP] Improving coverage in code

hbto committed:
Spanish Translation for l10n_pa_withholding (#1)

* [I18N] New Strings to translate

* Translated using Weblate (Spanish)

Currently translated at 0.0% (0 of 103 strings)

* Translated using Weblate (Spanish)

Currently translated at 100.0% (103 of 103 strings)

* Translated using Weblate (Spanish)

Currently translated at 100.0% (103 of 103 strings)

hbto commented:
@nhomar 
https://drive.google.com/file/d/0B1w9gokcJL9hWHlNRDMzajFyN2M/view

hbto commented:
Method to compute taxes needs to be fixed,
When wh_subject is `5` or `6` the withholding is performed based on
the whole invoice amount_total, rather than tax based.

hbto committed:
[FIX] Change business logic on how to fetch wh subject data and how to
condition the process to properly get the desired data

hbto committed:
[FIX] Business logic now takes into account withholdings from TDD/TDC.
[ADD] Added demo data to test Withholdings with TDD/TDC.
[IMP] UnitTesting: New business logic for withholding with TDD/TDC.
